design/proposals/ambient-namespaces.md

```diff
@@ -0,0 +1,52 @@
+# Ambient Namespaces
+
+Status: Pending
+
+Version: Alpha
+
+## Abstract
+
+We want to introduce functionality to enable the use of environment metadata (ex: cluster namespace, Server URI) in prototype and component manifest files.
+
+## Motivation
+
+* Environments have configured metadata, such as the cluster’s namespace or server URI. Users [have requested](https://github.com/ksonnet/ksonnet/issues/222) that these fields be referenceable in both prototypes and component files.
+* Component and prototype manifests should not reference a specific environment's details. See Goals and Non Goals for more clarity on what "specific" implies.
+
+Example:
+
+Consider the following [nginx prototype](https://github.com/ksonnet/parts/blob/9d78d6bb445d530d5b927656d2293d4f12654608/incubator/nginx/examples/nginx.jsonnet) where the namespace is currently hardcoded:
+
+```jsonnet
+local k = import 'k.libsonnet';
+local nginx = import '../nginx.libsonnet';
+
+local namespace = "dev-alex";
+local appName = "nginx-app";
+
+k.core.v1.list.new([
+  nginx.parts.deployment.withServerBlock(namespace, appName),
+  nginx.parts.service(namespace, appName),
+  nginx.parts.serverBlockConfigMap(namespace, appName),
+])
+```
+ 
+This example does not lend well to the ksonnet desired model where components are environment agnostic; meaning a component should not need information on which environment the resulting Kubernetes Objects will be deployed to. Yet, it is unavoidable that these component / prototype configurations may need to reference a namespace, server URI, or other metadata.
+
+## Goals
+
+* Encourage behavior for component and prototype manifests to be environment-agnostic. 
+  Example: It is encouraged to write `nginx.parts.service(env.namespace, appName)`, but not `local namespace = "foo"` or `if (env.namespace == "foo") ... else ...`.
```

Comment:
Interesting, why not something like `params.global.namespace` instead of `env.namespace`? Then you could put in the params file something like:

```
{
  global:: {
    namespace:: std.extVar("__ksonnet/global/namespace"),
  },
}
```

Or something.

This re-uses a bunch of constructs we already have, which is IMHO nicer.

Reply:
This is an interesting idea, and would work too. My only reservations (if that) would be:

`params.global.namespace` is a lot more wordy than `env.namespace` and does it articulate well enough that we are operating on an environment's attribute? That said, how opinionated do we want a ksonnet environment to be Ex: Since namespace is a metadata field of environment, should access / customization of a namespace be hidden by it's environment? If we begin to introduce them as global params, we seem to be introducing two areas where they can be configured. Not that this is necessarily bad, but I can see it causing confusion.

The plus side of this approach would allow for params to have fields such as:

```
"guestbook-ui": {
  namespace: global.namespace
}
```

Reply:
Those are good points. I think you're right that we should go with `env.*`, and you're right that this is likely more intuitive to most users.

But in that case, I'd add to the goals section: discoverability and auditing. By not using the `params` workflow, we can't do things like `ks params list` and `ks params diff`. It seems to me those are valuable, though, so it might be worth special casing them as part of the params workflow.

Reply:
Makes sense. I'll add that section along with an alternatives that details the params approach.